test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/TimeoutMock.scala

```diff
@@ -0,0 +1,110 @@
+package org.scalajs.testsuite.jsinterop
+
+import scala.annotation.tailrec
+import scala.collection.immutable.TreeMap
+import scala.scalajs.js
+import scala.scalajs.js.annotation.ScalaJSDefined
+
+object TimeoutMock {
+
+  import js.Dynamic.global
+
+  private val realSetTimeout =
+    global.setTimeout.asInstanceOf[js.Function]
+
+  private val realClearTimeout =
+    global.clearTimeout.asInstanceOf[js.Function]
+
+  private val mockSetTimeout: js.Function =
+    (fun: js.Function, delay: Int) => MockTimeouts.setTimeout(fun, delay)
+
+  private val mockClearTimeout: js.Function =
+    (timeout: MockTimeout) => MockTimeouts.clearTimeout(timeout)
+
+  private var installed = false
+
+  @noinline
+  def withMockedTimeout[A](body: (Int => Unit) => A): A = {
+    install()
+    try {
+      body(MockTimeouts.tick)
+    } finally {
+      uninstall()
+    }
+  }
+
+  private def install(): Unit = {
+    assert(!installed, "Mock timeout already installed.")
+    global.setTimeout = mockSetTimeout
+    global.clearTimeout = mockClearTimeout
+    installed = true
+  }
+
+  private def uninstall(): Unit = {
+    assert(installed, "Mock timeout is not installed.")
+    MockTimeouts.reinitialize()
+    global.setTimeout = realSetTimeout
+    global.clearTimeout = realClearTimeout
+    installed = false
+  }
+
+  private object MockTimeouts {
+    private var timeouts: TreeMap[Int, List[MockTimeout]] = _
+    private var currentTime: Int = _
+    reinitialize()
+
+    def reinitialize(): Unit = {
+      timeouts = TreeMap.empty
+      currentTime = 0
+    }
+
+    def setTimeout(fun: js.Function, delay: Int): MockTimeout = {
+      val timeout = new MockTimeout(fun)
+      assert(delay >= 0, "Delay should be positive.")
+      val triggerTime = currentTime + delay
+      assert(triggerTime >= 0, "Time overflow")
+      timeouts = timeouts.updated(triggerTime,
+        timeout :: timeouts.getOrElse(triggerTime, Nil))
+      timeout
+    }
+
+    def clearTimeout(timeout: MockTimeout): Unit = {
+      timeout.clearTimeout()
+    }
+
+    def tick(ms: Int): Unit = {
+      assert(installed, "`tick` should only be used inside withMockedTimeout { tick => ...}")
+      @tailrec
+      def tickLoop(ms: Int): Unit = {
+        assert(ms >= 0)
+        val targetTime = currentTime + ms
+        assert(targetTime >= 0)
+        timeouts.headOption match {
+          case Some((fTime, fTimeouts)) if fTime <= targetTime =>
+            currentTime = fTime
+            timeouts = timeouts.drop(1)
```

Comment:
This doesn't look like a very efficient way to deal with a `TreeMap`.

Have you considered the very simple implementation with a single `List[(Int, MockTiemout)]`:

Insert is:
```scala
val (before, after) = timeouts.span(_._1 <= triggerTime)
timeouts = before ::: timeout :: after
```
`tick` is:
```scala
val (before, after) = timeouts.span(_._1 <= targetTime)
timeouts = after // Note: this *must* come first
before.foreach(_.execute()))
```